src/ngraph/pass/nop_elimination.cpp

```diff
@@ -33,76 +38,156 @@
 using namespace std;
 using namespace ngraph;
 
-#define TI(x) std::type_index(typeid(x))
+#define TI(x) x::type_info
+
+static bool remove_update_name(const std::shared_ptr<Node>& node,
+                               const std::shared_ptr<Node>& node_input)
+{
+    bool has_result_output = false;
+    for (auto& output : node->output(0).get_target_inputs())
+    {
+        if (as_type<op::Result>(output.get_node()))
+        {
+            has_result_output = true;
+        }
+    }
+    // ignore trivial elimination
+    if (has_result_output && as_type_ptr<ngraph::op::Parameter>(node_input))
+    {
+        return false;
+    }
+    if (!has_result_output || node_input->get_users().size() == 1)
+    {
+        node_input->set_friendly_name(node->get_friendly_name());
+        replace_node(node, node_input);
+        return true;
+    }
+    return false;
+}
 
 static bool eliminate_pad(const std::shared_ptr<Node>& node)
 {
-    auto pad = std::static_pointer_cast<op::Pad>(node);
+    auto pad = std::static_pointer_cast<op::v0::Pad>(node);
+    // skip if shapes are dynamic
+    if (pad->get_input_partial_shape(0).is_dynamic() ||
+        pad->get_output_partial_shape(0).is_dynamic())
+    {
+        return false;
+    }
+
     if (pad->get_input_shape(0) == pad->get_output_shape(0))
     {
-        replace_node(node, node->get_argument(0));
-        return true;
+        return remove_update_name(node, node->get_argument(0));
     }
     return false;
 }
 
 static bool eliminate_sum(const std::shared_ptr<Node>& node)
 {
-    auto sum = std::static_pointer_cast<op::Sum>(node);
+    auto sum = std::static_pointer_cast<op::v0::Sum>(node);
     if (sum->get_reduction_axes().empty())
     {
-        replace_node(node, node->get_argument(0));
-        return true;
+        return remove_update_name(node, node->get_argument(0));
     }
     return false;
 }
 
 static bool eliminate_convert(const std::shared_ptr<Node>& node)
 {
-    auto convert = std::static_pointer_cast<op::Convert>(node);
-    if (convert->get_convert_element_type() == convert->get_argument(0)->get_element_type())
+    bool is_out_type_agnostic = false;
+    static const std::set<NodeTypeInfo> type_agnostic{TI(op::v3::NonZero)};
+    if (node->output(0).get_target_inputs().size() == 1)
     {
-        replace_node(node, node->get_argument(0));
-        return true;
+        auto& out = *node->output(0).get_target_inputs().begin();
+        is_out_type_agnostic = type_agnostic.count(out.get_node()->get_type_info()) == 1;
```

Comment:
for an expanding convert, why do we need to care if user node is type_agnostic? Is it because this is a nop_elimination pass? If we have convert(char) -> convert(int16) -> convert(int32)->{ any set of output ops}, can't we do away with the first 2 converts? For type_agnostic, any convert input node can be eliminated, if that's the only user of the convert?

Reply:
this change only handles convert cases where the following op accepts any type (eg: NonZero).

Reply:
I meant that, we can also add the case I mentioned above, here

Reply:
will be addressed in follow-up PR